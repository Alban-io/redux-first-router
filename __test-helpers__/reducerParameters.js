import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
