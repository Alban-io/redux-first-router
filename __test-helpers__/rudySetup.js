import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
