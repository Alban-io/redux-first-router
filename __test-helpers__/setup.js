import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
