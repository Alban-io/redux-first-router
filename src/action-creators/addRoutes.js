// @flow

import { ADD_ROUTES } from '../index'
import type { RoutesMap } from '../flow-types'

export default (routes: RoutesMap) => ({
  type: ADD_ROUTES,
  payload: { routes }
})
