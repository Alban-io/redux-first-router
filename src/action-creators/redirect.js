// @flow
import type { Action } from '../flow-types'

export default (action: Action, status: number = 302, committed: boolean = true) => {
  action.meta = action.meta || {}
  action.meta.location = action.meta.location || {}
  action.meta.location.kind = action.kind = 'redirect'
  action.meta.location.status = action.meta.location.status || status

  if (action.meta.location.committed === undefined) {
    action.meta.location.committed = committed // used to determine whether to do `history.redirect` or `history.push` (see utils/isRedirect.js)
  }

  return action
}
