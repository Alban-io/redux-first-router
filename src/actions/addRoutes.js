// @flow

import { ADD_ROUTES } from '../types'
import type { RoutesMap } from '../flow-types'

export default (routes: RoutesMap, formatRoute: ?Function) => ({
  type: ADD_ROUTES,
  payload: { routes, formatRoute }
})

// NOTE: see `src/utils/formatRoutes.js` for implemenation of corresponding pathlessRoute
