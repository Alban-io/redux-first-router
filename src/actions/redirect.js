// @flow
import type { Action } from '../flow-types'

export default (action: Action, status: number = 302) => ({
  ...action,
  location: {
    ...action.location,
    status,
    url: null, // insures action can go through pipeline (`url` indicates its passed through already), see `utils/isTransformed.js`
    kind: 'redirect'
  }
})
