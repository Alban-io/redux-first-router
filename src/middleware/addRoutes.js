import { ADD_ROUTES } from '../index'
import formatRoutes from '../utils/formatRoutes'

export default (api) => (req, next) => {
  if (req.action && req.action.type === ADD_ROUTES) {
    Object.assign(req.routes, formatRoutes(req.action.payload.routes, true))
    return req.commitDispatch(req.action)
  }

  return next()
}

