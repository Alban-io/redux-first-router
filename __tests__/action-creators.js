import redirect from '../src/action-creators/redirect'
import addRoutes from '../src/action-creators/addRoutes'

import { setupAll } from '../__test-helpers__/setup'

it('redirect(action) - sets action.meta.location.kind === "redirect"', () => {
  const receivedAction = { type: 'ANYTHING' }
  const action = redirect(receivedAction) /*? */

  expect(action.meta.location.kind).toEqual('redirect')
})

it('addRoutes(routes) - adds routes to routesMap', async () => {
  const newRoutes = {
    FOO: { path: '/foo' },
    BAR: { path: '/bar' }
  }

  const { store } = await setupAll()

  const action = addRoutes(newRoutes)
  await store.dispatch(action)

  expect(store.getState()).toMatchSnapshot()

  await store.dispatch({ type: 'FOO' })
  expect(store.getState().location.type).toEqual('FOO')
})
