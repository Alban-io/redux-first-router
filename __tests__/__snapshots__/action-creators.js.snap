// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`historyCreateAction() - returns action created when history/address_bar chanages 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": true,
      "current": Object {
        "pathname": "/info/foo",
        "payload": Object {
          "param": "foo",
        },
        "type": "INFO_PARAM",
      },
      "history": Object {
        "entries": Array [
          "/",
        ],
        "index": 0,
        "length": 1,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "/prev",
        "payload": Object {},
        "type": "PREV",
      },
      "redirect": undefined,
    },
  },
  "payload": Object {
    "param": "foo",
  },
  "type": "INFO_PARAM",
}
`;

exports[`middlewareCreateAction() - [action not matched to any routePath] 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": undefined,
      "current": Object {
        "pathname": "/prev",
        "payload": Object {
          "someKey": "foo",
        },
        "type": "@@pure-redux-router/NOT_FOUND",
      },
      "history": Object {
        "entries": Array [
          "/",
          "/prev",
        ],
        "index": 1,
        "length": 2,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "/prev",
        "payload": Object {},
        "type": "PREV",
      },
      "redirect": undefined,
    },
  },
  "payload": Object {
    "someKey": "foo",
  },
  "type": "@@pure-redux-router/NOT_FOUND",
}
`;

exports[`middlewareCreateAction() - returns action created when middleware detects connected/matched action.type 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": undefined,
      "current": Object {
        "pathname": "/info/foo",
        "payload": Object {
          "param": "foo",
        },
        "type": "INFO_PARAM",
      },
      "history": Object {
        "entries": Array [
          "/",
          "/info/foo",
        ],
        "index": 1,
        "length": 2,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "/prev",
        "payload": Object {},
        "type": "PREV",
      },
      "redirect": undefined,
    },
  },
  "payload": Object {
    "param": "foo",
  },
  "type": "INFO_PARAM",
}
`;
